### ISSUE TITLE
TLS/DTLS: OCSP Stapling support

### ISSUE THREAD

**Issue by RonEld** [ARMmbed/mbedtls#880]
Add support for OCSP.
Referenced in [rfc6066](https://tools.ietf.org/html/rfc6066#section-8) and [rfc2560](https://tools.ietf.org/html/rfc2560)

**Comment by RonEld** [comment]
Internal ref: IOTSSL-144

**Comment by JanneKiiskila** [comment]
Internal ref: IOTPREQ-172.

**Comment by igaztanaga** [comment]
Any news on about this feature?

**Comment by VitSolcik** [comment]
Hi, Is OCSP in any road map or planned? Thanks for info!

**Comment by gilles-peskine-arm** [comment]
Hi @VitSolcik, you can see our roadmap at https://developer.trustedfirmware.org/w/mbed-tls/roadmap/ . Unfortunately OCSP stapling is not in the plans for the next few quarters.

We (the Arm Mbed TLS team) did some internal work on the topic in 2017. I don't know how much was done. If someone is interested in picking it up, ping us and we'll see what we can find from back then.

**Comment by VitSolcik** [comment]
Than you for quick response! One of our client is interested in this. We will ping you if needed.

**Comment by soldatkin** [comment]
Hello,
Could you tell me if ocsp stapling feature is  implemented in mbedtls?

**Comment by gilles-peskine-arm** [comment]
No, unfortunately we still haven't found the time to implement it, and it's still not on our [roadmap](https://mbed-tls.readthedocs.io/en/latest/project/roadmap/).

**Comment by razvanphp** [comment]
Hello,
We are working on porting mbedtls & microcontroller support back into [EVerest project](https://github.com/EVerest/libiso15118/commit/9b77420b1a16ab6cf32c7526dce090bb19ba8eb0).
Since the ISO15118 standard for EV chargers & cars mandates OCSP stapling, we want to contribute on adding this. Can you publish the WIP internal branch so we can work on it further?
Thank you!

**Comment by gilles-peskine-arm** [comment]
@razvanphp I've pushed the 2017–2018 work to https://github.com/Mbed-TLS/mbedtls/tree/dev/andresag01/iotssl-1673-ocsp-resp-verify .

This is an old branch but you may use it under the current license of Mbed TLS.

This code is old and will need some work to adapt it to current Mbed TLS versions. It is incomplete. Here's what the author (who no longer works on Mbed TLS or for Arm) wrote about it at the time:

> This is the first draft PR of the OCSP response verification. The verification of OCSP responses is described in RFC 6960.
>
> This PR includes the main code for the verification together with several unit tests in test_suite_x509parse_ocsp. The API aims to mirror the functionality present in existing X509 modules such as CRT and CRL. However, several features present there have been omitted at this stage. Furthermore, as a first draft, I expect that reviewers will have comments regarding the API/features/etc so I have left on purpose a couple of untested bits of code which are tedious to create, such as tests for the info functions (but please feel free to point out if you see untested code).
>
> I have tried to split the code into several self-contained commits to make it easier to review, but I realise that this is a lot of code. So I have also annotated the code extensively with comments with detailed descriptions as well as TODO notes to emphasise the some features are missing or (IMHO) the RFC was not clear on what to do.
>
> Finally, please note that I have tried to reuse the existing APIs and code from the other X509 modules. This means that in some places I have had to expose code as internal functions which were previously only visible to a module (such as x509_crt.c). In other places, I have had to extend slightly the existing APIs to fit new purposes. I expect there will be contention on some of these, but please remember that this is only a draft and I am happy to hear other ideas on how the code can be improved.

Since then we never had time to review or continue the work.

Please note that in the short or even medium term, we have no bandwidth to help with the work or review it. So it's probably better than nothing if you want to make fork of the library, but we know that we cannot accept a contribution of this size in the near future.
